<!DOCTYPE html>
<html>
<head>
  <title>GENE315 - Genomes</title>
  <meta charset="utf-8">
  <meta name="description" content="GENE315 - Genomes">
  <meta name="author" content="Prof Tony Merriman &amp; A/Prof Mik Black">
  <meta name="generator" content="slidify" />
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta http-equiv="X-UA-Compatible" content="chrome=1">
  <link rel="stylesheet" href="libraries/frameworks/io2012/css/default.css" media="all" >
  <link rel="stylesheet" href="libraries/frameworks/io2012/css/phone.css" 
    media="only screen and (max-device-width: 480px)" >
  <link rel="stylesheet" href="libraries/frameworks/io2012/css/slidify.css" >
  <link rel="stylesheet" href="libraries/highlighters/highlight.js/css/tomorrow.css" />
  <base target="_blank"> <!-- This amazingness opens all links in a new tab. -->  
  
  <!-- Grab CDN jQuery, fall back to local if offline -->
  <script src="http://ajax.aspnetcdn.com/ajax/jQuery/jquery-1.7.min.js"></script>
  <script>window.jQuery || document.write('<script src="libraries/widgets/quiz/js/jquery.js"><\/script>')</script> 
  <script data-main="libraries/frameworks/io2012/js/slides" 
    src="libraries/frameworks/io2012/js/require-1.0.8.min.js">
  </script>
  
  

</head>
<body style="opacity: 0">
  <slides class="layout-widescreen">
    
    <!-- LOGO SLIDE -->
        <slide class="title-slide segue nobackground">
  <hgroup class="auto-fadein">
    <h1>GENE315 - Genomes</h1>
    <h2>Lecture 9 - Tools for identifying variation between genomes</h2>
    <p>Prof Tony Merriman &amp; A/Prof Mik Black<br/>Department of Biochemistry, University of Otago</p>
  </hgroup>
  <article></article>  
</slide>
    

    <!-- SLIDES -->
    <slide class="" id="slide-1" style="background:;">
  <article data-timings="">
    <style>
.title-slide {
  background-color: #FFFFFF; /* #CBE7A5; #EDE0CF; ; #CA9F9D*/
}

.title-slide hgroup > h1{
  font-family: 'Oswald', sans-serif;
  color: #000000;
}

.title-slide hgroup > h2{
  font-family: 'Signika Negative', 'Calibri', sans-serif;
  color: #000000;
}

/* Fonts and Spacing */
article p, article li, article li.build{
  font-family: 'Helvetica', 'Crimson Text', 'Garamond',  'Palatino', sans-serif;
  text-align: justify;
  font-size:30px;
  line-height: 1.1em;
}

slide:not(.segue) h2{
  font-family: 'Signika Negative', 'Calibri', Arial, sans-serif;
  font-size: 52px;
  font-style: normal;
  font-weight: bold;
  text-transform: normal;
  letter-spacing: -2px;
  line-height: 1.2em;
/*  color: #193441;*/
/*  color: #02574D;*/
  color: #CC2904;
}

slide:not(.segue) h3{
  font-family: 'Helvetica', 'Crimson Text', 'Garamond',  'Palatino', sans-serif;
  font-size: 22px;
  font-style: normal;
  line-height: 1.5em;
}

slide:not(.segue) h4{
#  font-family: 'Signika Negative', 'Calibri', Arial, sans-serif;
  font-family: 'Calibri', Arial, sans-serif;
  font-size: 40px;
  font-style: normal;
  font-weight: bold;
  text-transform: normal;
  letter-spacing: -2px;
  line-height: 1.3em;
/*  color: #193441;*/
  color: #02574D;
/*  color: #CC2904;*/
}

</style>

<h2>Sequencing data</h2>

<ul>
<li>Next (Second) Generation Sequencing has revolutionized how we study genetic variation.</li>
<li>It is now routine to generate a whole-genome characterisation of SNPs (single nucleotide polymorphisms), InDels (Insertions/Deletions) and 
SVs (structural variants, e.g., copy number variation) for an individual. </li>
</ul>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-2" style="background:;">
  <hgroup>
    <h2>Recap - generating NGS data</h2>
  </hgroup>
  <article data-timings="">
    <p><center><img src="PNG/ngs1.png" height="500"></center></p>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-3" style="background:;">
  <hgroup>
    <h2>Data format: FASTQ</h2>
  </hgroup>
  <article data-timings="">
    <ul>
<li>The read data generated by the sequencer are stored in the <em>FASTQ</em> format.</li>
<li>FASTQ allows the storage of both sequence and quality information for each read.</li>
<li>This is a compact text-based format that has become the de facto standard for storing data from next generation sequencing experiments.</li>
</ul>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-4" style="background:;">
  <hgroup>
    <h2>Fastq format</h2>
  </hgroup>
  <article data-timings="">
    <p><img src="PNG/fastq1.png" height="455"></p>

<h3><a href="http://en.wikipedia.org/wiki/FASTQ_format">http://en.wikipedia.org/wiki/FASTQ_format</a></h3>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-5" style="background:;">
  <hgroup>
    <h2>Fastq format</h2>
  </hgroup>
  <article data-timings="">
    <p><img src="PNG/fastq2.png" height="475"></p>

<h3><a href="http://en.wikipedia.org/wiki/FASTQ_format">http://en.wikipedia.org/wiki/FASTQ_format</a></h3>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-6" style="background:;">
  <hgroup>
    <h2>Identifying variation</h2>
  </hgroup>
  <article data-timings="">
    <ul>
<li>In situations where a reference genome is available (i.e., organisms for
which the genome has previously been sequenced and assembled), the sequencing of additional members of the species is often refered to as <em>resequencing</em>.</li>
<li>In order to identify variation in the resequenced samples, the sequence data is compared to the existing reference genome to allow differences to be identified.</li>
<li>This involves finding the place in the reference genome that each read matches to.</li>
<li>Due to high sequence similarity within members of the same species, most reads should map to the reference.</li>
</ul>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-7" style="background:;">
  <hgroup>
    <h2>Mapping reads to the reference genome</h2>
  </hgroup>
  <article data-timings="">
    <p><center><img src="PNG/ngs2.png" height="460"></center></p>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-8" style="background:;">
  <hgroup>
    <h2>Tools for generating alignments</h2>
  </hgroup>
  <article data-timings="">
    <ul>
<li>There are MANY software packages available for aligning data from next generation sequencing experiments.</li>
<li>Two of the most popular are:

<ul>
<li>BWA:  <a href="http://bio-bwa.sourceforge.net">http://bio-bwa.sourceforge.net</a></li>
<li>Bowtie:  <a href="http://bowtie-bio.sourceforge.net">http://bowtie-bio.sourceforge.net</a></li>
</ul></li>
<li>Both utilize the “Burrows-Wheeler Transform.”</li>
</ul>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-9" style="background:;">
  <article data-timings="">
    <p><center><img src="PNG/trapnell.png" height="500"></center></p>

<h3>Trapnell, C., &amp; Salzberg, S. L. (2009). How to map billions of short reads onto genomes. Nature Biotechnology, 27(5), 455–457.</h3>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-10" style="background:;">
  <hgroup>
    <h2>Alignment formats</h2>
  </hgroup>
  <article data-timings="">
    <ul>
<li>SAM (Sequence Alignment/Map) format has become the <em>de facto</em> standard for storing alignment data.</li>
<li>BAM is a binary version of SAM allowing more efficient storage.</li>
</ul>

<p><center><img src="PNG/samtools.png" height="300"></center></p>

<h3>Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., et al. (2009). The Sequence Alignment/Map format and SAMtools. Bioinformatics (Oxford, England), 25(16), 2078–2079.</h3>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-11" style="background:;">
  <hgroup>
    <h2>Aligned data - finding SNPs</h2>
  </hgroup>
  <article data-timings="">
    <p><center><img src="PNG/ngs3.png" height="240"></center></p>

<ul>
<li>Alignment to the reference genome allows identification of single base 
differences (i.e., SNPs) in a new sample.</li>
<li>There are <strong>MANY</strong> software tools that can perform SNP detection.</li>
<li>A reasonable level of coverage is required for accurate detection, particularly for 
heterozygotes.</li>
</ul>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-12" style="background:;">
  <hgroup>
    <h2>Using read depth to find copy number changes</h2>
  </hgroup>
  <article data-timings="">
    <ul>
<li>Sequence data can also be used to identify changes in copy number for specific regions of DNA.</li>
<li>Various methods exist, but a relatively simple approach involves identifying changes in read depth that are consistent with an increase or decrease in DNA copy number.</li>
</ul>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-13" style="background:;">
  <hgroup>
    <h2>Detecting DNA duplication</h2>
  </hgroup>
  <article data-timings="">
    <p><center><img src="PNG/ngs4.png" height="500"></center></p>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-14" style="background:;">
  <hgroup>
    <h2>Detecting DNA deletion</h2>
  </hgroup>
  <article data-timings="">
    <p><center><img src="PNG/ngs5.png" height="500"></center></p>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-15" style="background:;">
  <hgroup>
    <h2>Quantifying read depth across region</h2>
  </hgroup>
  <article data-timings="">
    <p><center><img src="PNG/ngs6.png" height="300"></center></p>

<ul>
<li>Break the region of interest into evenly sized windows, and count the number of reads
that start in each window.</li>
<li>Use a statistical algorithm to identify &quot;change points&quot; - places where the average 
number of reads per window changes (up or down), over a substantial distance.</li>
</ul>

  </article>
  <!-- Presenter Notes -->
</slide>

    <slide class="backdrop"></slide>
  </slides>
  <div class="pagination pagination-small" id='io2012-ptoc' style="display:none;">
    <ul>
      <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=1 title='NA'>
         1
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=2 title='Recap - generating NGS data'>
         2
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=3 title='Data format: FASTQ'>
         3
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=4 title='Fastq format'>
         4
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=5 title='Fastq format'>
         5
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=6 title='Identifying variation'>
         6
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=7 title='Mapping reads to the reference genome'>
         7
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=8 title='Tools for generating alignments'>
         8
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=9 title='NA'>
         9
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=10 title='Alignment formats'>
         10
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=11 title='Aligned data - finding SNPs'>
         11
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=12 title='Using read depth to find copy number changes'>
         12
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=13 title='Detecting DNA duplication'>
         13
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=14 title='Detecting DNA deletion'>
         14
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=15 title='Quantifying read depth across region'>
         15
      </a>
    </li>
  </ul>
  </div>  <!--[if IE]>
    <script 
      src="http://ajax.googleapis.com/ajax/libs/chrome-frame/1/CFInstall.min.js">  
    </script>
    <script>CFInstall.check({mode: 'overlay'});</script>
  <![endif]-->
</body>
  <!-- Load Javascripts for Widgets -->
  
  <!-- LOAD HIGHLIGHTER JS FILES -->
  <script src="libraries/highlighters/highlight.js/highlight.pack.js"></script>
  <script>hljs.initHighlightingOnLoad();</script>
  <!-- DONE LOADING HIGHLIGHTER JS FILES -->
   
  </html>